hardware/chisel/src/main/scala/core/TensorAlu.scala

```diff
@@ -97,38 +97,330 @@ class AluVector(implicit p: Parameters) extends Module {
+
+  val sIdle::sRun::sWait::Nil = Enum(3)
+  val state = RegInit(init=sIdle)
+  val inflight = RegInit(0.U(inflightBits.W))
+
+  val index_generator = Module(new TensorAluIndexGenerator)
+  val aluDataReadPipeDelay = 0 // available for pipelining
+
+  // State Machine for compute io.done correctly
+  io.done := false.B
+  when( state === sIdle && io.start) {
+    state := sRun
+  }.elsewhen( state === sRun && index_generator.io.last) {
+    state := sWait
+  }.elsewhen( state === sWait && inflight === 0.U) {
+    state := sIdle
+    io.done := true.B
+  }
+
+  index_generator.io.start := io.start
+  index_generator.io.dec := io.dec
+
+  // second term works around funny clearing in uop register file flopped output
+  io.uop.idx.valid := index_generator.io.valid || index_generator.io.src_valid
+  io.uop.idx.bits := index_generator.io.uop_idx
+
+  val valid_001 = ShiftRegister( index_generator.io.valid, aluDataReadPipeDelay + 1, resetData=false.B, en = true.B)
+  val valid_002 = RegNext( valid_001, init=false.B)
+  val valid_003 = RegNext( valid_002, init=false.B)
+  val valid_004 = RegNext( valid_003, init=false.B)
+
+  when( index_generator.io.valid && valid_004) {
+  }.elsewhen( index_generator.io.valid) {
+    assert( inflight =/= ((1<<inflightBits)-1).U)
+    inflight := inflight + 1.U
+  }.elsewhen( valid_004) {
+    assert( inflight =/= 0.U)
+    inflight := inflight - 1.U
+  }
+  when( state === sIdle) {
+    assert( inflight === 0.U)
+    inflight := 0.U
+  }
+
+  val src_valid_001 = ShiftRegister(
+    index_generator.io.src_valid,
+    aluDataReadPipeDelay + 1,
+    resetData=false.B, en = true.B)
+  val src_valid_002 = RegNext( src_valid_001, init=false.B)
+  val src_valid_003 = RegNext( src_valid_002, init=false.B)
+  val src_valid_004 = RegNext( src_valid_003, init=false.B)
+
+  val dst_idx_001 = ShiftRegister( index_generator.io.dst_idx, aluDataReadPipeDelay + 1)
+  val src_idx_001 = ShiftRegister( index_generator.io.src_idx, aluDataReadPipeDelay + 1)
+
+  val uop_data_001 = ShiftRegister(io.uop.data, aluDataReadPipeDelay)
+
+  val dst_offset = uop_data_001.bits.u0
+
+  val w = dst_offset.getWidth
+  val u2 = uop_data_001.bits.u2.asTypeOf(UInt(w.W))
+  val s = log2Ceil(p(CoreKey).inpMemDepth)
+  val u1 = uop_data_001.bits.u1.asTypeOf(UInt(w.W))
+  val src_offset = (u2 << s) | u1
+
+  // split registers of stage 2 by data groups
+  //val accRdIdxValid = valid_002 || src_valid_002
+  val accRdIdxValid = valid_001 || src_valid_001
+  for (idx <- 0 until dataSplitFactor) {
+    //io.acc.rd(idx).idx.valid := accRdIdxValid
+    io.acc.rd(idx).idx.valid := RegNext(accRdIdxValid)
+  }
+
+  val new_src_idx_001 = src_idx_001 + src_offset
+  val src_idx_002 = RegNext( new_src_idx_001)
+  val src_idx_003 = RegNext( src_idx_002)
+
+  val new_dst_idx_001 = dst_idx_001 + dst_offset
+  val dst_idx_002 = RegNext( new_dst_idx_001)
+  val dst_idx_003 = RegNext( dst_idx_002)
+  val dst_idx_004 = RegNext( dst_idx_003)
```

Comment:
what would you think on using `_r1` instead of `_001` or some other format that starts with a letter  followed by a number

Reply:
Changed to _r# throughout